Stack/Opc.Ua.Core/Security/Certificates/CertificateValidator.cs

```diff
@@ -851,19 +851,55 @@ protected virtual async Task InternalValidate(X509Certificate2Collection certifi
 
             if ((certificateKeyUsage & X509KeyUsageFlags.DataEncipherment) == 0)
             {
-                throw new ServiceResultException(StatusCodes.BadCertificateUseNotAllowed, "Usage of certificate is not allowed.");
+                sresult = new ServiceResult(StatusCodes.BadCertificateUseNotAllowed,
+                    null,
+                    null,                    
+                    "Usage of certificate is not allowed.",
+                    null,
+                    sresult
+                    );
             }
 
             // check if minimum requirements are met
             if (m_rejectSHA1SignedCertificates && IsSHA1SignatureAlgorithm(certificate.SignatureAlgorithm))
             {
-                throw new ServiceResultException(StatusCodes.BadCertificatePolicyCheckFailed, "SHA1 signed certificates are not trusted");
+                sresult = new ServiceResult(StatusCodes.BadCertificatePolicyCheckFailed,
+                    null,
+                    null,
+                    "SHA1 signed certificates are not trusted",
+                    null,
+                    sresult
+                    );
             }
 
             int keySize = CertificateFactory.GetRSAPublicKeySize(certificate);
             if (keySize < m_minimumCertificateKeySize)
             {
-                throw new ServiceResultException(StatusCodes.BadCertificatePolicyCheckFailed, "Certificate doesn't meet minimum key length requirement");
+                sresult = new ServiceResult(StatusCodes.BadCertificatePolicyCheckFailed,
+                    null,
+                    null,
+                    "Certificate doesn't meet minimum key length requirement",
+                    null,
+                    sresult
+                    );
+            }
+
+            if (issuedByCA && chainIncomplete)
+            {
+                sresult = new ServiceResult(StatusCodes.BadCertificateChainIncomplete,
```

Comment:
@ThomasNehring, have you looked into an implementation which chains the ServiceResultException instead of ServiceResult?
I think its just a gut feeling, but it feels more natural to get chained exceptions than a single exception with chained serviceresult? Thoughts?